File: javascript/14.longest-common-prefix.js
/*
 * @lc app=leetcode id=14 lang=javascript
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
  let prefix = "";

  // Solution 1: two dimensional array like
  if (strs.length === 1) return strs[0];
  let i = 0,
    j = 0;
  while (j < strs.length - 1) {
    if (i >= strs[j].length || strs[j][i] !== strs[j + 1][i]) break;
    if (j === strs.length - 2) {
      prefix += strs[j][i++];
      j = 0;
    } else j++;
  }

  // Solution 2: same as Solution 1 but more concise
  // for (let i = 0; i < strs[0].length; i++) {
  //   if (strs.every((str) => str[i] === strs[0][i])) prefix += strs[0][i];
  //   else break;
  // }

  return prefix;
};
// @lc code=end

File: rust/14.longest-common-prefix.rs
/*
 * @lc app=leetcode id=14 lang=rust
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        return match strs.is_empty() {
            true => "".to_string(),
            false => strs.iter().skip(1).fold(strs[0].clone(), |acc, x| {
                acc.chars()
                    .zip(x.chars())
                    .take_while(|(a, b)| a == b)
                    .map(|(x, _)| x)
                    .collect()
            }),
        }
    }
}
// @lc code=end
